File: EasyDD - pillarcompression/displacement.m
function [utilda] = displacement(p,segments,loop_list,nu)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Code to calculate the displacement vector at a point p 
% ue to a dislocation structure (closed loops / networks)
%
% Steve Fitzgerald, 16 Oct 2013
% Francesco Ferroni, 28 Oct 2013
%
% University of Oxford, Department of Materials, Parks Road, United Kingdom
% Defects Group & Materials for Fission and Fusion Power
%
% p: field point [x y z]
% u: displacement [ux uy uz]
% tol: tolerance to check whether node in loop is coplanar (normal distance
% to plane, in units of b). 5-10b would be acceptable.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

utilda = [0 0 0]; %initialize
con1 = (1-2*nu)/(8*pi*(1-nu));
con2 = 1/(8*pi*(1-nu));

n_loops = max(loop_list(:,2)); 
size_list = size(loop_list,1);
size_seglist = size(segments,1);

for z=1:n_loops
    
    %isolate single "loop" from "loop_list"
    index=false(size_list,1);
    for d=1:size_list
        if z==loop_list(d,2);
            index(d)=true;
        end
    end
    loop=loop_list(index,1);
    n=size(loop,1);
       
    %generate node pairs in loop
    id_pairs = zeros(n,2);
    for i=1:n-1
        id_pairs(i,1) = loop(i);
        id_pairs(i,2) = loop(i+1);
    end
    id_pairs(n,1) = loop(n);
    id_pairs(n,2) = loop(1);
    
    index=false(size_seglist,1);
    for d=1:size_seglist
        for i=1:n 
            if id_pairs(i,1)==segments(d,1) && id_pairs(i,2)==segments(d,2)
                index(d)=true;
            elseif id_pairs(i,1)==segments(d,2) && id_pairs(i,2)==segments(d,1)
                index(d)=true;
            end
        end
    end
    %make 'nodes' and 'segs' list
    nodes=segments(index,6:8);
    b=segments(index,3:5);
    segs=segments(index,9:11)-segments(index,6:8);
    
    %check burgers consistency. they should be all the same since they are
    %in a single loop, I think.
    for d=1:n
        if not ( all(b(:,1) == b(1,1)) && all(b(:,2) == b(1,2)) && all(b(:,3) == b(1,3)) )
            disp('Not-identical Burgers vector for loop number ',num2str(z))
            pause;
        end
    end
    b=b(1,:);
    
    %calculate loop plane normal [a b c d], (using 2 segments of loop)
    vec_a = segs(1,:)/norm(segs(1,:));
    vec_b = -segs(end,:)/norm(segs(end,:));
    plane_n = cross(vec_a,vec_b);
    d = -dot(nodes(1,:),plane_n);
    plane_n(4) = d;
    
%     %check if loop nodes are coplanar. not necessary: no cross-slip.
%     cros
%     vrts=rn(loop(:),1:3);
%     coplanar = vrts(:,1)*plane_n(1) + vrts(:,2)*plane_n(2) + vrts(:,3)*plane_n(3) + plane_n(4);
%     test=any(coplanar>tol | coplanar<-tol);
%     if test
%         disp('Loop is not coplanar')
%     end
    
    
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% SPF

    %centroid
    C = sum(nodes)./n;

    % t vectors (unit tangents to segments)
    t = zeros(size(nodes));
    for i=1:n
        t(i,:)=segs(i,:)./norm(segs(i,:));
    end

    % R vectors (from P to nodes)
    R = zeros(size(nodes));
    for i=1:n
        R(i,:) = nodes(i,:) - p;
    end
    RC = C - p;

    % absolute values
    modR = zeros(n,1);
    for i=1:n
        modR(i) = norm(R(i,:));
    end
    modRC = norm(RC);

    % lambda vectors (unit vectors of the R's)
    lam = zeros(size(nodes));
    for i=1:n
        lam(i,:)=R(i,:)./modR(i); %if point p lies on the lam vector, this creates NaN term! - FF
    end
    lamC = RC./modRC;

    % extra segments FROM nodes TO centroid
    extras = zeros(size(nodes));
    for i=1:n
        extras(i,:) = C - nodes(i,:);
    end

    % tangents for these -- FROM nodes TO centroid
    tex = zeros(size(nodes));
    for i=1:n
        tex(i,:) = extras(i,:)./norm(extras(i,:));
    end

    % displacement vector
    u = [0 0 0];

    % start from first node, make a triangle ABC

    for i=1:n-1;
        ab = t(i,:);
        bc = tex(i+1,:);
        ca = -tex(i,:);%opposite sense
        lamA = lam(i,:);
        lamB = lam(i+1,:);
        % already know lamC
        RA = modR(i);
        RB = modR(i+1);
        % already know modRC

        % get the three f's
        f = fab(b, ab, lamA, lamB, RA, RB) ...
            + fab(b, bc, lamB, lamC, RB, modRC) ...
            + fab(b, ca, lamC, lamA, modRC, RA);

        % and the three g's
        g = gab(b, lamA, lamB) + gab(b, lamB, lamC) + gab(b, lamC, lamA);

        % update displacement inc. solid angle
        u = u - b.*(solang(lamA, lamB, lamC , p , plane_n))./(4*pi) - con1.*f + con2.*g;
     end

        % last one
        ab = t(n,:);
        bc = tex(1,:);
        ca = -tex(n,:);%opposite sense
        lamA = lam(n,:);
        lamB = lam(1,:);
        % already know lamC
        RA = modR(n);
        RB = modR(1);
        % already know modRC

        % get the three f's
        f = fab(b, ab, lamA, lamB, RA, RB) ...
            + fab(b, bc, lamB, lamC, RB, modRC) ...
            + fab(b, ca, lamC, lamA, modRC, RA);

        % and the three g's
        g = gab(b, lamA, lamB) + gab(b, lamB, lamC) + gab(b, lamC, lamA);

        % update displacement inc. solid angle
        u = u - b.*(solang(lamA, lamB, lamC , p , plane_n))./(4*pi) - con1.*f + con2.*g;
        
        % add onto final displacement
        utilda = utilda + u;
        
         %disp('Oosterom Method (new)');
         %disp(solang2(lamA, lamB, lamC , p , plane_n));
       
         %disp('Huilier Method (original)');
         %disp(solang(lamA, lamB, lamC , p , plane_n));
end

utilda = utilda'; %make it a column vector

end

function [ out ] = fab( b, t, lamA, lamB, RA, RB )
%calculates f vector

numerator = RB*(1 + dot(lamB,t));
denominator = RA*(1 + dot(lamA,t));

logarithm = log(numerator/denominator);

out = cross(b,t).*logarithm;

end

function [ out ] = gab( b, lamA, lamB )
%calculates g vector

numerator = dot(b,cross(lamA,lamB)).*(lamA + lamB);
denominator = 1 + dot(lamA,lamB);

out = numerator./denominator;

end

function [ omega ] = solang( lamA, lamB, lamC , p , plane_n )
%calculates solid angle due to triangular loop ABC
% 1) this may not be robust, check different formulae for the 
% spherical excess (wiki spherical geometry)
% 2) it could be a lot faster to calculate a,b,c in the main loop 
% for each pair of lambdas, then use a,b,c as input arguments 
a = acos(dot(lamB,lamC));
b = acos(dot(lamC,lamA));
c = acos(dot(lamA,lamB));

s = (a + b + c)/2;

svec = [s s-a s-b s-c]./2;

temp = tan(svec(1))*tan(svec(2))*tan(svec(3))*tan(svec(4));

omega = 4*atan(sqrt(abs(temp)));

if dot(lamA,plane_n(1:3)) > 0
    sign = 1;
else
    sign = -1;
end

omega = -sign*omega;

%Eq(18) Barnett 1985
%test whether point is above or below plane or on it.
% x = dot(p,plane_n(1:3)) + plane_n(4);
% if x < 0
%     omega = -omega;
% end

end

function [ omega ] = solang2( lamA, lamB, lamC , p , plane_n )
%alternative solid angle formula to see if edge dislocation miscalculation
%is stemming from some error in solang function.

%given 3 input vectors: lamA, lamB, lamC

determ = det([lamA;lamB;lamC]);
 
al = norm(lamA);
bl = norm(lamB);
cl = norm(lamC);

div = al*bl*cl + dot(lamA,lamB)*cl + dot(lamA,lamC)*bl + dot(lamB,lamC)*al;
at = atan2(determ, div);

if at<0
    at = at + pi;
end

omega = 2*at;

if dot(lamA,plane_n(1:3)) > 0
    sign = 1;
else
    sign = -1;
end

omega = -sign*omega;

%Eq(18) Barnett 1985
%test whether point is above or below plane
% x = dot(p,plane_n(1:3)) + plane_n(4);
% if x < 0
%     omega = -omega;
% end

end

File: EasyDD - pillarcompression/displacement.c
#define _USE_MATH_DEFINES
#include <math.h>
#include <stdio.h>

void triangleDisplacement(double *point, double *A, double *B, double *C, double *b, double *n, double NU, double *utilda);

/* Declare auxiliary functions */
void fab(double *b, double *t, double *lamA, double *lamB, double RA, double RB, double *f_vec);
void gab(double *b, double *lamA, double *lamB, double *g_vec);
void solang(double *lamA, double *lamB, double *lamC, double *p, double *plane_n, double *omega);

/************************** MEX gateway function ***********************/

int  main() {
    
    double point[] = {0,0,0};
    double A[] = {10,0,-100};
    double B[] = {10,0,100};
    double C[] = {10,10,0};
    double b[] = {0.5,1,1};
    double n[] = {-1,0,0};
    double utilda[3];
    double NU=0.305;
    
    triangleDisplacement(point,A,B,C,b,n,NU,utilda);
    
    //for (int i=0; i<3; i++){
    //    printf("%f \n",utilda[i]);
    //}

return 0;
}

/**************************************************************************/

void triangleDisplacement(double *point, double *A, double *B, double *C, double *b, double *plane_n, double NU, double *utilda)
{
    const double C1 = (1.0-2.0*NU)/(8.0*M_PI*(1.0-NU));
    const double C2 = 1.0/(8.0*M_PI*(1.0-NU));
    
    /*     a -----> b
            ^      /
             \    /
              \  /
               c          */
    
    /* A is p1, B is p2, C is centroid */
    
    double RA[3], RB[3], RC[3];
    double vecAB[3], vecBC[3], vecCA[3];
    for (int i=0; i<3; i++){
        RA[i] = A[i] - point[i];
        RB[i] = B[i] - point[i];
        RC[i] = C[i] - point[i];
        
        vecAB[i] = B[i] - A[i];
        vecBC[i] = C[i] - B[i];
        vecCA[i] = A[i] - C[i];
    }
    
    const double modABinv = 1.0/sqrt(vecAB[0]*vecAB[0] + vecAB[1]*vecAB[1] + vecAB[2]*vecAB[2]);
    const double modBCinv = 1.0/sqrt(vecBC[0]*vecBC[0] + vecBC[1]*vecBC[1] + vecBC[2]*vecBC[2]);
    const double modCAinv = 1.0/sqrt(vecCA[0]*vecCA[0] + vecCA[1]*vecCA[1] + vecCA[2]*vecCA[2]);
    
    const double modRA = sqrt(RA[0]*RA[0] + RA[1]*RA[1] + RA[2]*RA[2]);
    const double modRB = sqrt(RB[0]*RB[0] + RB[1]*RB[1] + RB[2]*RB[2]);
    const double modRC = sqrt(RC[0]*RC[0] + RC[1]*RC[1] + RC[2]*RC[2]);
    const double modRAinv = 1.0/modRA;
    const double modRBinv = 1.0/modRB;
    const double modRCinv = 1.0/modRC;
    
    /*unit tangents along the directed segments AB, BC, and CA*/
    double tAB[3], tBC[3], tCA[3];
    double lamA[3], lamB[3], lamC[3];
    for (int i=0; i<3; i++){
        tAB[i] = vecAB[i]*modABinv;
        tBC[i] = vecBC[i]*modBCinv;
        tCA[i] = vecCA[i]*modCAinv;
        
        lamA[i] = RA[i]*modRAinv;
        lamB[i] = RB[i]*modRBinv;
        lamC[i] = RC[i]*modRCinv;
    }
    
//     printf("lamA \n");
//     for (int i=0; i<3; i++){
//         printf("%f \n",lamA[i]);
//     }
//     printf("lamB \n");
//     for (int i=0; i<3; i++){
//         printf("%f \n",lamB[i]);
//     }
//     printf("lamC \n");
//     for (int i=0; i<3; i++){
//         printf("%f \n",lamC[i]);
//     }
    
    /*calculate fAB, fBC, and fCA*/
    double fAB[3], fBC[3], fCA[3];
    fab(b,tAB,lamA,lamB,modRA,modRB,fAB);
    fab(b,tBC,lamB,lamC,modRB,modRC,fBC);
    fab(b,tCA,lamC,lamA,modRC,modRA,fCA);
    
//     printf("fAB \n");
//     for (int i=0; i<3; i++){
//         printf("%f \n",fAB[i]);
//     }
//     printf("fBC \n");
//     for (int i=0; i<3; i++){
//         printf("%f \n",fBC[i]);
//     }
//     printf("fCA \n");
//     for (int i=0; i<3; i++){
//         printf("%f \n",fCA[i]);
//     }
    
    /*calculate gAB, gBC, and gCA*/
    double gAB[3], gBC[3], gCA[3];
    gab(b,lamA,lamB,gAB);
    gab(b,lamB,lamC,gBC);
    gab(b,lamC,lamA,gCA);
    
//     printf("gAB \n");
//     for (int i=0; i<3; i++){
//         printf("%f \n",gAB[i]);
//     }
//     printf("gBC \n");
//     for (int i=0; i<3; i++){
//         printf("%f \n",gBC[i]);
//     }
//     printf("gCA \n");
//     for (int i=0; i<3; i++){
//         printf("%f \n",gCA[i]);
//     }
    
    /*calculate solid angle*/
    double omega[1];
    solang(lamA,lamB,lamC,point,plane_n,omega);
    
    //printf("Solid Angle \n");
    //printf("%f \n",omega[0]);
    
    for (int i=0; i<3; i++){
        utilda[i] = -b[i]*omega[0]/(4.0*M_PI) - C1*(fAB[i] + fBC[i] + fCA[i]) + C2*(gAB[i] + gBC[i] + gCA[i]);
    }
    
}

/*****************************Auxiliary functions******************************/

/*f vector calculation*/
void fab(double *b, double *t, double *lamA, double *lamB, double RA, double RB, double *f_vec)
{
    const double numerator = RB*(1.0 + lamB[0]*t[0] + lamB[1]*t[1] + lamB[2]*t[2]);
    const double denominator = RA*(1.0 + lamA[0]*t[0] + lamA[1]*t[1] + lamA[2]*t[2]);
    const double logarithm = log(numerator/denominator);
    
    f_vec[0] = logarithm*(b[1]*t[2] - b[2]*t[1]);
    f_vec[1] = logarithm*(b[2]*t[0] - b[0]*t[2]);
    f_vec[2] = logarithm*(b[0]*t[1] - b[1]*t[0]);

}

/*g vector calculation*/
void gab(double *b, double *lamA, double *lamB, double *g_vec)
{
    const double invdenom = 1.0/(1.0 + (lamA[0]*lamB[0] + lamA[1]*lamB[1] + lamA[2]*lamB[2]));
    
    double lamAxlamB[3];
    lamAxlamB[0] = lamA[1]*lamB[2] - lamA[2]*lamB[1];
    lamAxlamB[1] = lamA[2]*lamB[0] - lamA[0]*lamB[2];
    lamAxlamB[2] = lamA[0]*lamB[1] - lamA[1]*lamB[0];
    
    const double bdlamAxlamB = b[0]*lamAxlamB[0] + b[1]*lamAxlamB[1] + b[2]*lamAxlamB[2];
    
    double numerator[3];
    numerator[0] = bdlamAxlamB*(lamA[0] + lamB[0]);
    numerator[1] = bdlamAxlamB*(lamA[1] + lamB[1]);
    numerator[2] = bdlamAxlamB*(lamA[2] + lamB[2]);
    
    g_vec[0] = numerator[0]*invdenom;
    g_vec[1] = numerator[1]*invdenom;
    g_vec[2] = numerator[2]*invdenom;
    
}

/*solid angle calculation*/
void solang(double *lamA, double *lamB, double *lamC, double *p, double *plane_n, double *omega)
{
    /*determinant of [lamA;lamB;lamC]
     | lamA[1] lamA[2] lamA[3] |
     | lamB[1] lamB[2] lamB[3] |
     | lamC[1] lamC[2] lamC[3] | */
    const double cross1 = lamB[1]*lamC[2] - lamC[1]*lamB[2];
    const double cross2 = lamB[0]*lamC[2] - lamC[0]*lamB[2];
    const double cross3 = lamB[0]*lamC[1] - lamC[0]*lamB[1];
    const double detLam = lamA[0]*cross1 + lamA[1]*cross2 + lamA[2]*cross3;
    
    /*magnitude of lamA, lamB, lamC*/
    const double al = sqrt(lamA[0]*lamA[0] + lamA[1]*lamA[1] + lamA[2]*lamA[2]);
    const double bl = sqrt(lamB[0]*lamB[0] + lamB[1]*lamB[1] + lamB[2]*lamB[2]);
    const double cl = sqrt(lamC[0]*lamC[0] + lamC[1]*lamC[1] + lamC[2]*lamC[2]);
    
    /*dot(A,B) , dot(A,C) , dot(B,C)*/
    const double AdotB = lamA[0]*lamB[0] + lamA[1]*lamB[1] + lamA[2]*lamB[2];
    const double AdotC = lamA[0]*lamC[0] + lamA[1]*lamC[1] + lamA[2]*lamC[2];
    const double BdotC = lamB[0]*lamC[0] + lamB[1]*lamC[1] + lamB[2]*lamC[2];
    
    /*main formula*/
    const double div = al*bl*cl + AdotB*cl + AdotC*bl + BdotC*al;
    double at = atan(div/detLam);
    
    if (at<0.0)
        at = at + M_PI;
    
    omega[0] = 2.0*at;
    
    /*Test whether point is below or above plane!*/
    const double x = p[0]*plane_n[0] + p[1]*plane_n[1] + p[2]*plane_n[2] + plane_n[3];
    if (x<0.0)
        omega[0] = -omega[0];
}
